src/main/java/gov/cabinetofice/gapuserservice/service/OneLoginService.java

```diff
@@ -282,6 +295,52 @@ public void migrateUser(final User user, final String jwt) {
                 .block();
     }
 
+    public void validateIdToken(IdTokenDto decodedIdToken) {
+
+        long currentEpochSeconds = Instant.now().getEpochSecond();
+
+        if (!decodedIdToken.getIss().equals(oneLoginBaseUrl.concat("/")) ||
+                !decodedIdToken.getAud().equals(clientId) ||
+                currentEpochSeconds > decodedIdToken.getExp() ||
+                currentEpochSeconds < decodedIdToken.getIat()) {
+            log.error("Invalid id token {}", decodedIdToken);
+            throw new UnauthorizedException("Invalid id token");
+        }
+    }
+
+    public void validateUserSub(String IdTokenSub, String userInfoStub) {
+        if (!IdTokenSub.equals(userInfoStub)) {
+            log.error("User subs do not match {} {}", IdTokenSub, userInfoStub);
+            throw new UnauthorizedException("Invalid User sub");
+        }
+    }
+
+
+    public void validateAuthTokenSignatureAndAlgorithm(String authToken) {
+
+        try {
+            SignedJWT signedAuthToken = SignedJWT.parse(authToken);
+            JWSAlgorithm jwtAlgorithm = JWSAlgorithm.parse(signedAuthToken.getHeader().getAlgorithm().getName());
+            String keyId = signedAuthToken.getHeader().getKeyID();
+
+            JWKSet jwkSet = JWKSet.load(new URL(oneLoginBaseUrl.concat("/.well-known/jwks.json")));
+            Optional<JWK> optionalMatchingJwk = Optional.ofNullable(jwkSet.getKeyByKeyId(keyId));
+            JWK matchingJwk = optionalMatchingJwk.orElseThrow(() -> new UnauthorizedException
+                    ("Matching JWK not found for key ID: " + keyId));
+
+            ECDSAVerifier verifier = new ECDSAVerifier((ECKey) matchingJwk);
+
+            if (!signedAuthToken.verify(verifier) || !signedAuthToken.getHeader().getAlgorithm().equals(jwtAlgorithm)) {
```

Comment:
invert these? no point trying to verify if the algorithm is wrong (and verification will always fail)

Reply:
oh actually is this ever going to fail?

```js
signedAuthToken.getHeader().getAlgorithm().equals(jwtAlgorithm)
```

we pull the algorithm out of the jwt header on line 323 so isn't this check is essentially circular?

Reply:
Yeah you're right, need to compare the Algorithm matches the JWKSet instead I think

Reply:
yeah